Simplify network-adjusted time warning logic

stickies-v opened the pull request:
[An earlier approach](https://github.com/bitcoin/bitcoin/commits/1d226ae1f984c5c808f5c24c431b959cdefa692e/) in #28956 involved simplifying and refactoring the network-adjusted time calculation logic, but this was eventually [left out](https://github.com/bitcoin/bitcoin/pull/28956#issuecomment-1904214370) of the PR to make it easier for reviewers to focus on consensus logic changes.

Since network-adjusted time is now only used for warning/informational purposes, cleaning up the logic (building on @dergoegge's approach in #28956) should be quite straightforward and uncontroversial. The main changes are:

- Previously, we would only calculate the time offset from the first 199 outbound peers that we connected to. This limitation is now removed, and we have a proper rolling calculation. I've reduced the set to 50 outbound peers, which seems plenty.
- Previously, we would automatically use the network-adjusted time if the difference was < 70 mins, and warn the user if the difference was larger than that. Since there is no longer any automated time adjustment, I've changed the warning threshold to ~~20~~ 10 minutes (which is an arbitrary number).
- Previously, a warning would only be raised once, and then never again until node restart. This behaviour is now updated to  1) warn to log for every new outbound peer for as long as we appear out of sync, 2) have the RPC warning toggled on/off whenever we go in/out of sync, and 3) have the GUI warn whenever we are out of sync (again), but limited to 1 messagebox per 60 minutes
- no more globals
- remove the `-maxtimeadjustment` startup arg

Closes #4521

dergoegge committed:
[net processing] Move nTimeOffset to net_processing

dergoegge committed:
[net processing] Introduce PeerManagerInfo

For querying statistics/info from PeerManager. The median outbound time
offset is the only initial field.

dergoegge committed:
[net processing] Move IgnoresIncomingTxs to PeerManagerInfo

stickies-v committed:
[net processing] Use std::chrono for type-safe time offsets

stickies-v committed:
Remove timedata

With the introduction and usage of TimeOffsets, there is no more need
for this file. Remove it.

stickies-v committed:
Add TimeOffsets helper class

This helper class is an alternative to CMedianFilter, but without a
lot of the special logic and exceptions that we needed while it was
still used for consensus.

dergoegge commented:
Concept ACK

Thanks for picking this up!

dergoegge commented on src/test/fuzz/timedata.cpp:
```diff
@@ -1,31 +0,0 @@
-// Copyright (c) 2020-2021 The Bitcoin Core developers
```
Could you add a fuzz test for `TimeOffsets` to replace the timedata one?

stickies-v replied:
Can do, but given how simple the interface is I think it wouldn't be anything more than the below (untested)? Could e.g. add the `.Size()` method back but adding it just for fuzzing might not be ideal?

<details>
<summary>git diff on 55b476b015</summary>

```diff
diff --git a/src/Makefile.test.include b/src/Makefile.test.include
index c77056f29c..c20b531032 100644
--- a/src/Makefile.test.include
+++ b/src/Makefile.test.include
@@ -381,6 +381,7 @@ test_fuzz_fuzz_SOURCES = \
  test/fuzz/string.cpp \
  test/fuzz/strprintf.cpp \
  test/fuzz/system.cpp \
+ test/fuzz/timeoffsets.cpp \
  test/fuzz/torcontrol.cpp \
  test/fuzz/transaction.cpp \
  test/fuzz/tx_in.cpp \
diff --git a/src/test/fuzz/timeoffsets.cpp b/src/test/fuzz/timeoffsets.cpp
new file mode 100644
index 0000000000..f0d2a8f739
--- /dev/null
+++ b/src/test/fuzz/timeoffsets.cpp
@@ -0,0 +1,22 @@
+// Copyright (c) 2020-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <test/fuzz/FuzzedDataProvider.h>
+#include <test/fuzz/fuzz.h>
+#include <test/fuzz/util.h>
+#include <node/timeoffsets.h>
+
+#include <chrono>
+
+
+FUZZ_TARGET(timeoffsets)
+{
+    FuzzedDataProvider fuzzed_data_provider(buffer.data(), buffer.size());
+    TimeOffsets offsets{};
+    while (fuzzed_data_provider.remaining_bytes() > 0) {
+        (void)offsets.Median();
+        offsets.Add(std::chrono::seconds{fuzzed_data_provider.ConsumeIntegral<int64_t>()});
+        offsets.WarnIfOutOfSync();
+    }
+}

```
</details>

dergoegge replied:
> but given how simple the interface is I think it wouldn't be anything more than the below

That's good enough. We just want the code to be under test, in case we have change it in the future.

stickies-v replied:
Done. I was a bit worried `WarnIfOutOfSync()` outputting to stderr would mess with the fuzzer (as [here](https://github.com/bitcoin/bitcoin/pull/29623/commits/48cc160d70e4d918f5e2e5440b0ab5891499f5cf)), but running it locally seems to work fine so far, not sure if the appropriate paths have been hit already though.

Resolving for now.

stickies-v commented:
Force pushed to:
- address a [thread safety analysis issue](https://github.com/bitcoin/bitcoin/pull/29623/checks?check_run_id=22527737631) by annotating `WarnIfOutOfSync()` with a `EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)` 
- fix [failing functional test](https://github.com/bitcoin/bitcoin/pull/29623/checks?check_run_id=22527737628) by resetting node_miner's mocktime before connecting to the next outbound
- rebase to latest master to fix native macos job

alfonsoromanz commented:
Code review ACK 55b476b015d3a79ec5ea176247d2a70163cd034b

I also:
- Ran the unit test, which is passing
- Ran the functional test, which is passing
- Verified that the peer window now shows the offset of each peer.

<img width="300" alt="Screenshot 2024-03-12 at 11 36 48" src="https://github.com/bitcoin/bitcoin/assets/19962151/7cd58278-085c-4047-b43f-e71b538096f3">
<img width="303" alt="Screenshot 2024-03-12 at 11 37 20" src="https://github.com/bitcoin/bitcoin/assets/19962151/84801775-c730-4f03-bbd2-87571124a73d">
<img width="250" alt="Screenshot 2024-03-12 at 11 06 07" src="https://github.com/bitcoin/bitcoin/assets/19962151/32f3256e-fb57-4b69-af4b-97ecdfee9017">

stickies-v commented:
Force pushed to:
- remove dependency on C++20's `std::ranges::sort` which is [not yet supported](https://github.com/bitcoin/bitcoin/pull/29623/checks?check_run_id=22558502170) by the `no wallet, libbitcoinkernel` ci task
- add `timeoffsets` fuzz target as per [review comment](https://github.com/bitcoin/bitcoin/pull/29623#discussion_r1521434495)
- improve includes

> Verified that the peer window now shows the offset of each peer.

FYI this is not new functionality, you should see the same behaviour on `master`.

vasild commented:
Concept ACK, I love simplifications or refactors that are net removing code (+213 −363).

ariard commented:
> Previously, we would only calculate the time offset from the first 199 outbound peers that we connected to. This limitation is now removed, and we have a proper rolling calculation. I've reduced the set to 50 outbound peers, which seems plenty.
Previously, we would automatically use the network-adjusted time if the difference was < 70 mins, and warn the user if the difference was larger than that. Since there is no longer any automated time adjustment, I've changed the warning threshold to 20 minutes (which is an arbitrary number).

on the 2 value selections (50 outbound peers):
- 50 sounds reasonable, to be fair best would be to add timestamp in `m_outbound_time_offsets` from `FEELER` connection, it’s an outbound connection type per se. shuffling things more at the AS-level.
- 20 min, i can suggest to pick up 10 min to match average interval-block-time of ~10 min

maflcko reviewed (commented):
left a nit, might review later

stickies-v commented on src/test/fuzz/timeoffsets.cpp:
```diff
@@ -0,0 +1,22 @@
+#include <chrono>
+
+
+FUZZ_TARGET(timeoffsets)
```
Done.

stickies-v replied:
[Reverted back to `timeoffsets`](https://github.com/bitcoin/bitcoin/pull/29623#issuecomment-2004000023)

stickies-v commented on src/rpc/net.cpp:
```diff
@@ -679,9 +678,10 @@ static RPCHelpMan getnetworkinfo()
-        obj.pushKV("localrelay", !node.peerman->IgnoresIncomingTxs());
+        auto peerman_info{node.peerman->GetInfo()};
+        obj.pushKV("localrelay", !peerman_info.ignores_incoming_txs);
+        obj.pushKV("timeoffset", peerman_info.median_outbound_time_offset.count());
```
Done

stickies-v reviewed (commented):
Thank you for the reviews everyone, I've force-pushed to address all comments (and no other changes).

> 50 sounds reasonable, to be fair best would be to add timestamp in m_outbound_time_offsets from FEELER connection, it’s an outbound connection type per se. shuffling things more at the AS-level.

That's already happening, we're considering all types of outbound connections (including feelers)

> 20 min, i can suggest to pick up 10 min to match average interval-block-time of ~10 min

What's your rationale for matching it with the average block-time interval? I don't see how they're related. I'm not opposed to a shorter threshold (like 10 minutes), just want to avoid warnings occurring too frequently so users start to get used to and ignore them. ~~Want to avoid bikeshedding over this, so going to keep as is for now until more reviewers voice their preference.~~ _edit: updated threshold to 10 minutes_

stickies-v commented:
Force-pushed to revert the [fuzz harness name change](https://github.com/bitcoin/bitcoin/pull/29623#discussion_r1528075169), because for a couple of commits, we now have two harnesses with the same name - until the old one is removed in `Remove timedata` commit. I'll open a PR in `qa-assets` to rename the input folder (thx dergoegge for the suggestion).

dergoegge reviewed (approved):
Code review ACK a42d9b5041cdc522e8a4b81944e45b7f30458c26

vasild commented on src/warnings.cpp:
```diff
@@ -56,6 +62,11 @@ bilingual_str GetWarnings(bool verbose)
         warnings_verbose.emplace_back(warnings_concise);
     }
 
+    if (g_timeoffset_warning) {
+        warnings_concise = _("Your computer's date and time appear out of sync with the network, this may lead to consensus failure. Please ensure it is correct.");
+        warnings_verbose.emplace_back(warnings_concise);
+    }
```
Once `g_timeoffset_warning` is set to `true` it will never be set back to `false` and this will warning will persist even after the problem is resolved?

stickies-v replied:
Correct, as is roughly the current behaviour, as per https://github.com/bitcoin/bitcoin/pull/29623#discussion_r1531234359

vasild commented on test/functional/feature_maxtipage.py:
```diff
@@ -43,6 +43,10 @@ def test_maxtipage(self, maxtipage, set_parameter=True, test_deltas=True):
 
+        # reset time to system time so we don't have a time offset with the ibd node the next
+        # time we connect to it, ensuring TimeOffsets::WarnIfOutOfSync() doesn't output to stderr
+        node_miner.setmocktime(0)
```
Could squash commits:
587349072f `[net processing] Use TimeOffsets to warn when node is out of sync`
d790e776ca `[test] reset miner mocktime in maxtipage to prevent time offset warning`

vasild commented on src/timedata.cpp:
```diff
@@ -1,116 +0,0 @@
-    g_time_offsets.input(nOffsetSample);
-    LogPrint(BCLog::NET, "added time data, samples %d, offset %+d (%+d minutes)\n", g_time_offsets.size(), nOffsetSample, nOffsetSample / 60);
-
-    // There is a known issue here (see issue #4521):
-    //
-    // - The structure g_time_offsets contains up to 200 elements, after which
-    // any new element added to it will not increase its size, replacing the
-    // oldest element.
-    //
-    // - The condition to update nTimeOffset includes checking whether the
-    // number of elements in g_time_offsets is odd, which will never happen after
-    // there are 200 elements.
-    //
-    // But in this case the 'bug' is protective against some attacks, and may
-    // actually explain why we've never seen attacks which manipulate the
-    // clock offset.
-    //
-    // So we should hold off on fixing this and clean it up as part of
-    // a timing cleanup that strengthens it in a number of other ways.
```
https://github.com/bitcoin/bitcoin/pull/4526#issuecomment-49115517
I wonder what those "attacks" are? @gmaxwell

stickies-v replied:
I don't know myself, but I'd argue it's probably not relevant anymore as consensus dependencies on network-adjusted time have been removed

vasild replied:
Right, it dawned on me that previously we would assume the network's time, ie "adjust" our time according to the network and this has been removed, so "the network" can no longer mess with our clock. Sorry for the noise.

stickies-v commented on src/node/timeoffsets.h:
```diff
@@ -0,0 +1,39 @@
+    static constexpr size_t N{50};
+    //! Minimum difference between system and network time for a warning to be raised.
+    static constexpr std::chrono::minutes m_warn_threshold{10};
+    mutable std::atomic<bool> m_warning_emitted{false};  // Avoid flooding the user with warnings.
```
For GUI and RPC warnings, this PR pretty much preserves the current behaviour: GUI messagebox gets shown only once (until restarted), and the RPC will keep firing forever (on master, it will be reset when another `SetMiscWarning()` is called, but in practice I don't think that really happens frequently).

For the debug log, however, this PR quite significantly improves things by
1) doing an actual rolling calculation, which means that if clocks start drifting, the user will eventually still get a warning, whereas on master the time adjustment system completely stops after the first 199 outbound handshakes
2) producing repeated log warnings (one for every outbound handshake) for as long _and only for as long_ as the median time offset exceeds the threshold.

While I agree that for GUI & RPC things should be improved, this PR is no regression, and I'd rather it be done in a separate PR. Avoiding flooding seems less trivial to me than adding a simple timer: having a GUI message popup every 10 minutes is quite annoying if someone decides to flood the network with nodes with wrong clocks.

> Is there a way for the user to see the median network time, as perceived by his node?
yes, through `getnetworkinfo` a user can see the median offset

> 1 month later the issues resurface - there is no warning or indication about it?
there will be warning logs every time we handshake with an outbound peer and the median timeoffset exceeds the threshold

stickies-v replied:
See https://github.com/stickies-v/bitcoin/commit/0146553aaab53b6092bb54e0b2c71282fef44787 for a potential implementation on turning {GUI, RPC} warnings off again when the offsets fall within the threshold again. Happy to open a follow-up PR with this if you like the approach, I think this PR is too close to merge to overhaul it again as per my previous comment, what do you think?

stickies-v commented on src/node/timeoffsets.cpp:
```diff
@@ -0,0 +1,57 @@
+    auto median{std::max(Median(), std::chrono::seconds(std::numeric_limits<int64_t>::min() + 1))};
+    if (std::chrono::abs(median) <= m_warn_threshold) return false;
+
+    bilingual_str msg{_("Your computer's date and time appear out of sync with the network, "
+                        "this may lead to consensus failure. Please ensure it is correct.")};
```
I don't immediately see what practical benefit it would offer to the user (but I'm happy to add it if I need to re-touch.

vasild replied:
"Your computer's clock is 765 seconds behind the network. The difference is more than 600 seconds and this may lead to a consensus failure. Please ensure your computer's clock is correct."

vasild replied:
Consider this ^ while retouching. I think it is more useful because it tells the user how wrong is the clock and what are the expectations of Bitcoin Core. Otherwise (to the user) it is not clear whether Bitcoin Core is too picky and whether this is about +/- 1 second or something like that. Referring the user to the `getnetworkinfo` RPC may also be useful.

vasild reviewed (approved):
ACK a42d9b5041cdc522e8a4b81944e45b7f30458c26

stickies-v commented:
While [working on the fuzz inputs](https://github.com/bitcoin-core/qa-assets/pull/178#issuecomment-2012479018), I noticed that the memory usage for this new `timeoffsets` harness is unusually high. Doing one more small force-push to patch the fuzz and unit test to:
- disable logging in the fuzz test, which was causing the memory leak (thanks @dergoegge)
- use `noui.h` to prevent the GUI from flooding the fuzz and unit test with (expected) warnings

It's a pretty small test-only diff, so I'm hoping reviewers can give this a quick re-review? Thank you!

`git range-diff a42d9b5041cdc522e8a4b81944e45b7f30458c26...3dfdb22c1da687bd86e3325ec02c7b38e15895b4`

stickies-v commented:
Changed to draft until I've resolved the merge conflict with #29279 which relies on setmocktime

maflcko reviewed (commented):
lgtm 3dfdb22c1da687bd86e3325ec02c7b38e15895b4 🌽

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: lgtm 3dfdb22c1da687bd86e3325ec02c7b38e15895b4 🌽
3k5YN4x2rtvpoxcx88mjfPhCYoWnOCqHeZLoKsYaWrghV1HDaqNUzr4RiQ0kTXkWkdr/HPRYmcI77mwR2WiBBg==
```

</details>

stickies-v reviewed (commented):
Thanks for the in-depth review everyone.

Force-pushed to rebase to make CI green again (#29704) and address all outstanding review comments, given that this has now become a larger push anyway. The main changes are:
- RPC warnings are properly reset when the issue is resolved, GUI warnings can now be shown more than once (e.g. when the node falls out of sync again), but only once per 60 minutes to prevent flooding.
- the warning message has been improved to make it more clear what the user has to/can do
- re-added debug log whenever an offset is added (i.e. an outbound peer connects)

vasild commented on src/node/timeoffsets.cpp:
```diff
@@ -0,0 +1,57 @@
+
+    bilingual_str msg{_("Your computer's date and time appear out of sync with the network, "
+                        "this may lead to consensus failure. Please ensure it is correct.")};
+    LogWarning("%s\n", msg.original);
```
Grepping in my node's log for how frequent outbound connections are made - it would take about 1.5 days to make 200 new outbound connections. Since this is about the median, I guess 101 new connections after fixing the clock will be enough for us to stop the warning. That is like 20ish hours. Quite long. Can we do better?

Currently we only save the difference between our time and the peer's time at the start of the connection and we have no clue that this difference has changed due to our clock being adjusted/corrected. If we save the value of the system_clock and steady_clock at the start of the connection, in addition to the peer's offset, at a later time we can deduce that our clock has been adjusted and add that diff/adjustment to the peer's offset to immediately figure out that "now our clock is in sync with that peer's clock".

So, instead of `peer_offset`, it would be a matter of using the following instead:
`peer_offset + steady_clock_diff - system_clock_diff`, assuming:
`steady_clock_diff = steady_clock_now - steady_clock_when_connected`
`system_clock_diff = system_clock_now - system_clock_when_connected`.

Example:
---
Lets say our clock is wrong, 15 minutes ahead. When we connect to a peer at `17:00:00` our time they send us a time that is 15 minutes ago and we save:
`peer_offset = -15min`,
`system_clock_when_connected = 17:00:00`,
`steady_clock_when_connected = 1000` (some obscure value [seconds]).
One hour later, our clock has been fixed/corrected, by moving it 15 min backwards (or stopping to move for 15 min). So,
`system_clock_now = 17:45:00`,
`steady_clock_now = 4600` (the obscure value 1000 plus 3600 because 1 hour has passed).
And then:
`steady_clock_diff = 4600 - 1000 = 60 min`
`system_clock_diff = 17:45:00 - 17:00:00 = 45 min`
The difference between those is 15 minutes and adding that to `peer_offset` gives us offset 0, meaning that now our clock is in sync with the peer's clock.

stickies-v replied:
Using both system and steady clock is an interesting idea and one I hadn't considered, thanks. I'm going to look into this a bit more and will follow up shortly.

> it would take about 1.5 days to make 200 new outbound connections

FYI this PR reduces 200 -> 50 outbounds so your calculations are a bit off.

achow101 commented:
ACK 2ef71c73582be554e565ada3f8a6ca77c2cd79f1

sr-gi commented on src/test/timeoffsets_tests.cpp:
```diff
@@ -0,0 +1,71 @@
+{
+    noui_test_redirect();  // prevent the TimeOffsets::WarnIfOutOfSync() GUI MessageBoxes from flooding the log
+    BOOST_CHECK(IsWarningRaised({{-60min, -40min, -30min, 0min, 10min}}));
+    BOOST_CHECK(IsWarningRaised({5, 21min}));
```
Since the warn threshold was reduced from 20 to 10, it may make more sense to change this to 11

sr-gi reviewed (changes requested):
ACK https://github.com/bitcoin/bitcoin/commit/2ef71c73582be554e565ada3f8a6ca77c2cd79f1

Left some comments inline, mostly nits/things that may need renaming. 

Also, `qt/guiutils.h` still refers to `CNodeCombinedStats.nTimeOffset` after 5084ad48462b3169c99a81aeb21a339c3a9a4728, in a couple of places, but it may be better to change that for `time_offset` given `nTimeOffset` is not a thing anymore.

maflcko reviewed (commented):
lgtm 2ef71c73582be554e565ada3f8a6ca77c2cd79f1   🛴

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: lgtm 2ef71c73582be554e565ada3f8a6ca77c2cd79f1   🛴
Q9sgCN5S+/6cPhk6lYNjDonkzAs6Gm8QI7XynAlXDMCM29pQnMy67XGFMTPuBfZCMwdyVCo/S1sFWJI8l5xkCw==
```

</details>

vasild reviewed (commented):
Approach ACK 2ef71c73582be554e565ada3f8a6ca77c2cd79f1

Wrt adjusting the peer's offset with local clock corrections in order to stop the warnings immediately after the local clock is fixed (instead of after 4-5h) https://github.com/bitcoin/bitcoin/pull/29623#discussion_r1544783265, consider this:

<details>
<summary>[patch] adjust peer time with local clock corrections (untested)</summary>

```diff
diff --git i/src/node/timeoffsets.cpp w/src/node/timeoffsets.cpp
index 8488e47ff5..4dce6f9edd 100644
--- i/src/node/timeoffsets.cpp
+++ w/src/node/timeoffsets.cpp
@@ -24,27 +24,31 @@ void TimeOffsets::Add(std::chrono::seconds offset)
 {
     LOCK(m_mutex);
 
     if (m_offsets.size() >= N) {
         m_offsets.pop_front();
     }
-    m_offsets.push_back(offset);
+    m_offsets.emplace_back(offset, std::chrono::steady_clock::now(), std::chrono::system_clock::now());
     LogDebug(BCLog::NET, "Added time offset %+ds, total samples %d\n",
              Ticks<std::chrono::seconds>(offset), m_offsets.size());
 }
 
 std::chrono::seconds TimeOffsets::Median() const
 {
     LOCK(m_mutex);
 
     // Only calculate the median if we have 5 or more offsets
     if (m_offsets.size() < 5) return 0s;
 
+    const auto steady_now = std::chrono::steady_clock::now();
+    const auto system_now = std::chrono::system_clock::now();
     auto sorted_copy = m_offsets;
-    std::sort(sorted_copy.begin(), sorted_copy.end());
-    return sorted_copy[m_offsets.size() / 2];  // approximate median is good enough, keep it simple
+    std::sort(sorted_copy.begin(), sorted_copy.end(), [&](const peer_time_data& a, const peer_time_data& b) {
+        return a.Offset(steady_now, system_now) < b.Offset(steady_now, system_now);
+    });
+    return sorted_copy[m_offsets.size() / 2].Offset(steady_now, system_now);  // approximate median is good enough, keep it simple
 }
 
 bool TimeOffsets::WarnIfOutOfSync() const
 {
     // when median == std::numeric_limits<int64_t>::min(), calling std::chrono::abs is UB
     auto median{std::max(Median(), std::chrono::seconds(std::numeric_limits<int64_t>::min() + 1))};
diff --git i/src/node/timeoffsets.h w/src/node/timeoffsets.h
index 8bccae20e8..1d7a9f0153 100644
--- i/src/node/timeoffsets.h
+++ w/src/node/timeoffsets.h
@@ -21,13 +21,29 @@ private:
     //! Log the last time a GUI warning was raised to avoid flooding user with them
     mutable std::optional<std::chrono::steady_clock::time_point> m_warning_emitted GUARDED_BY(m_mutex){};
     //! Minimum time to wait before raising a new GUI warning
     static constexpr std::chrono::minutes m_warning_wait{60};
 
     mutable Mutex m_mutex;
-    std::deque<std::chrono::seconds> m_offsets GUARDED_BY(m_mutex){};
+    struct peer_time_data {
+        //! The difference between peer's time and our time, at the time of the P2P handshake.
+        std::chrono::seconds handshake_offset;
+        // The following two are used to detect system clock corrections/adjustments.
+        std::chrono::time_point<std::chrono::steady_clock> handshake_steady;
+        std::chrono::time_point<std::chrono::system_clock> handshake_system;
+
+        /// Calculate the offset as of now, taking into account any system clock corrections that
+        /// have happened since the peer connected.
+        std::chrono::seconds Offset(std::chrono::time_point<std::chrono::steady_clock> steady_now,
+                                    std::chrono::time_point<std::chrono::system_clock> system_now) const {
+            return handshake_offset +
+                std::chrono::duration_cast<std::chrono::seconds>(steady_now - handshake_steady) -
+                std::chrono::duration_cast<std::chrono::seconds>(system_now - handshake_system);
+        }
+    };
+    std::deque<peer_time_data> m_offsets GUARDED_BY(m_mutex){};
 
 public:
     /** Add a new time offset sample. */
     void Add(std::chrono::seconds offset) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
 
     /** Compute and return the median of the collected time offset samples. */
```
</details>

glozow commented on src/net_processing.cpp:
```diff
@@ -753,6 +753,8 @@ class PeerManagerImpl final : public PeerManager
     /** Next time to check for stale tip */
     std::chrono::seconds m_stale_tip_check_time GUARDED_BY(cs_main){0s};
 
+    TimeOffsets m_outbound_time_offsets;
```
needs doc?

glozow commented on src/rpc/net.cpp:
```diff
@@ -681,8 +680,8 @@ static RPCHelpMan getnetworkinfo()
     if (node.peerman) {
         auto peerman_info{node.peerman->GetInfo()};
         obj.pushKV("localrelay", !peerman_info.ignores_incoming_txs);
+        obj.pushKV("timeoffset", Ticks<std::chrono::seconds>(peerman_info.median_outbound_time_offset));
```
Or its own rpc commit?

glozow commented on src/net_processing.cpp:
```diff
@@ -3666,12 +3671,11 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+        peer->m_time_offset = nTime - GetTime();
         if (!pfrom.IsInboundConn()) {
             // Don't use timedata samples from inbound peers to make it
             // harder for others to tamper with our adjusted time.
```
+1, was going to say this should be removed in 2ef71c73582be554e565ada3f8a6ca77c2cd79f1, or say we are avoiding false warnings instead of preventing tampering of adjusted time

glozow commented on src/node/timeoffsets.h:
```diff
@@ -0,0 +1,42 @@
+class TimeOffsets
+{
+private:
+    static constexpr size_t N{50};
```
nit: documentation please, e.g. number of elements to track

glozow reviewed (commented):
approach ACK

stickies-v commented:
Thank you very much @sr-gi @maflcko @vasild @glozow @mzumsande for your extensive review.

Force pushed to address all review comments:
- in GUI: display the warning along with other warnings, instead of in a messagebox. This prevents the warning from [blocking the `msghand` thread](https://github.com/bitcoin/bitcoin/pull/29623#discussion_r1551733252) (and is a nicer approach in general)
- various nits around improved documentation, variable naming, references, help messages, ...

I probably won't be addressing nits in this PR anymore to minimize review burden, so if there are any left (please keep posting them if you see any) I'll keep them for a follow-up.

---

> Also, qt/guiutils.h still refers to CNodeCombinedStats.nTimeOffset

References updated, thanks

> Wrt adjusting the peer's offset with local clock corrections in order to stop the warnings immediately after the local clock is fixed (instead of after 4-5h) https://github.com/bitcoin/bitcoin/pull/29623#discussion_r1544783265, consider this:

I like the idea of using `steady_clock` here, but I would prefer to do it in a follow-up if there's appetite for it. It's nice to have the warning disappear more quickly when the issue is resolved, but I think a downside is that the timeoffset peer stats in GUI and RPC will be inconsistent with the ones we use for the warning. I'd started implementing a `PeerClock` approach based on your suggestion, and I think it's quite elegant, but it's a lot more code overhaul so not really ideal for this PR. WIP commit at https://github.com/bitcoin/bitcoin/compare/master...stickies-v:bitcoin:2024-03/remove-timedata-keep-clock

stickies-v reviewed (commented):
Thank you very much @sr-gi @maflcko @vasild @glozow @mzumsande for your extensive review.

Force pushed to address all review comments:
- in GUI: display the warning along with other warnings, instead of in a messagebox. This prevents the warning from https://github.com/bitcoin/bitcoin/pull/29623#discussion_r1551733252 (and is a nicer approach in general)
- various nits around improved documentation, variable naming, references, help messages, ...
- I probably won't be addressing nits in this PR anymore to minimize review burden, so if there are any left (please keep posting them if you see any) I'll keep them for a follow-up.

> Also, qt/guiutils.h still refers to CNodeCombinedStats.nTimeOffset

References updated, thanks

> Wrt adjusting the peer's offset with local clock corrections in order to stop the warnings immediately after the local clock is fixed (instead of after 4-5h) https://github.com/bitcoin/bitcoin/pull/29623#discussion_r1544783265, consider this:

I like the idea of using steady_clock here, but I would prefer to do it in a follow-up if there's appetite for it. It's nice to have the warning disappear more quickly when the issue is resolved, but I think a downside is that the timeoffset peer stats in GUI and RPC will be inconsistent with the ones we use for the warning. I'd started implementing a PeerClock approach based on your suggestion, and I think it's quite elegant, but it's a lot more code overhaul so not really ideal for this PR. WIP commit at https://github.com/bitcoin/bitcoin/compare/master...stickies-v:bitcoin:2024-03/remove-timedata-keep-clock

sr-gi commented on src/net_processing.cpp:
```diff
@@ -392,7 +392,7 @@ struct Peer {
     /** Time offset computed during the version handshake based on the
      * timestamp the peer send in the version message. */
-    std::atomic<int64_t> m_time_offset{0};
+    std::atomic<std::chrono::seconds> m_time_offset{0s};
```
Oh, my point was that it could have also been defined as `0s` there too, but I won't insist

sr-gi commented on src/net_processing.cpp:
```diff
@@ -3666,12 +3681,12 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
-        pfrom.nTimeOffset = nTimeOffset;
+        peer->m_time_offset = NodeSeconds{std::chrono::seconds{nTime}} - Now<NodeSeconds>();
         if (!pfrom.IsInboundConn()) {
             // Don't use timedata samples from inbound peers to make it
```
I don't think this was covered (or it may have been regressed)

stickies-v replied:
Sorry, I think it must have regressed with [this update in the next line](https://github.com/bitcoin/bitcoin/compare/2ef71c73582be554e565ada3f8a6ca77c2cd79f1..5362db10ab6d1d6697807c8f8494a2fc5bf41ff8#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R3687) - will fix if I have to retouch.

sr-gi commented:
Re-ACK [c6be144](https://github.com/bitcoin/bitcoin/pull/29623/commits/c6be144c4b774a03a8bcab5a165768cf81e9b06b)

stickies-v commented on src/node/timeoffsets.cpp:
```diff
@@ -0,0 +1,76 @@
+#include <limits>
+#include <optional>
+
+using namespace std::chrono_literals;
```
We `#include <util/time.h>` which has `using namespace std::chrono_literals;` which is why it compiles. I thought we would prefer (IWYU-style) to be explicit about used namespaces, but upon further investigation the purpose of #20602 seems to have been to _avoid_ that, so I'll remove this if I re-touch.

achow101 commented:
reACK c6be144c4b774a03a8bcab5a165768cf81e9b06b

dergoegge reviewed (approved):
utACK c6be144c4b774a03a8bcab5a165768cf81e9b06b

hebasto commented:
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/186.
